lucille-core/src/main/java/com/kmwllc/lucille/stage/CopyFields.java

```diff
@@ -1,70 +1,102 @@
- *   fields for a 1-1 mapping of results or supply one destination field for all of the source fields to be mapped into.</li>
+ *   <li>fieldMapping (Map&lt;String, String&gt;) : A 1-1 mapping of source field names to destination field names.</li>
  *   <li>update_mode (String, Optional) : Determines how writing will be handling if the destination field is already populated. Can be
- *   'overwrite', 'append' or 'skip'. Defaults to 'overwrite'.</li>
+ *   'overwrite', 'append' or 'skip'. Defaults to 'overwrite'. Cannot be set with isNested.</li>
+ *   <li>isNested (Boolean, Optional) : Sets whether or not to treat the given field name as nested json or a literal field name.
+ *   It true, the field name will be split on '.' and each part will be treated as a level of nesting. Cannot be set with update_mode.
+ *   Defaults to false. </li>
  * </ul>
  */
 public class CopyFields extends Stage {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private static final ObjectMapper mapper = new ObjectMapper();
 
   public static final Spec SPEC = SpecBuilder.stage()
-      .requiredList("source", new TypeReference<List<String>>(){})
-      .requiredList("dest", new TypeReference<List<String>>(){})
-      .optionalString("update_mode").build();
+      .requiredParent("fieldMapping", new TypeReference<Map<String, String>>() {})
+      .optionalString("update_mode")
+      .optionalBoolean("isNested").build();
 
-  private final List<String> sourceFields;
-  private final List<String> destFields;
+  private final Map<String, Object> fieldMapping;
   private final UpdateMode updateMode;
+  private final boolean isNested;
+  private Map<String, String[]> sourceToDestParts = new HashMap<>();
 
   public CopyFields(Config config) {
     super(config);
-    this.sourceFields = config.getStringList("source");
-    this.destFields = config.getStringList("dest");
+    this.fieldMapping = config.getConfig("fieldMapping").root().unwrapped();
     this.updateMode = UpdateMode.fromConfig(config);
+    this.isNested = ConfigUtils.getOrDefault(config, "isNested", false);
   }
 
   @Override
   public void start() throws StageException {
-    StageUtils.validateFieldNumNotZero(sourceFields, "Copy Fields");
-    StageUtils.validateFieldNumNotZero(destFields, "Copy Fields");
-    StageUtils.validateFieldNumsSeveralToOne(sourceFields, destFields, "Copy Fields");
+    if (this.fieldMapping.size() == 0) {
+      throw new StageException("fieldMapping must have at least one source-dest pair for Copy Fields");
+    }
+
+    if (this.isNested && this.updateMode != null) {
+      log.info("Cannot set both isNested and update_mode in Copy Fields at the same time. Ignoring update_mode, fields will be overwritten.");
+    }
+
+    // create set of destination field parts so we don't have to split them up for every doc
+    for (Entry<String, Object> fieldPair : this.fieldMapping.entrySet()) {
```

Comment:
suggest that when handling the nested case, we should iterate over a list of Pair<String[], String[]> where both the source and dest have been pre-split
when handling the no nested case, we can just iterate over the entries in the fieldMapping.entrySet()
this way, handling nested, we never have to split source or dest inside processDocument(), and when handling non-nested, we never have to join source or dest inside processDocument()

will share a code example separately